macros/src/parser/lifetimes.rs

```diff
@@ -0,0 +1,98 @@
+use quote::quote;
+use std::ops::Deref;
+use syn::{parse, spanned::Spanned, GenericArgument, Lifetime, PathArguments, Type};
+
+#[derive(Debug, Clone)]
+pub struct Lifetimes(Vec<Lifetime>);
+
+impl Lifetimes {
+    pub fn new() -> Lifetimes {
+        Lifetimes(Vec::new())
+    }
+
+    pub fn from_type(data_type: &Type) -> Result<Lifetimes, parse::Error> {
+        let mut lifetimes = Lifetimes::new();
+        get_lifetimes(data_type, &mut lifetimes)?;
+        Ok(lifetimes)
+    }
+
+    pub fn push(&mut self, lifetime: &Lifetime) {
+        if !self.0.contains(lifetime) {
+            self.0.push(lifetime.to_owned());
+        }
+    }
```

Comment:
`push()` generally isn't used as a conditional. Perhaps some other nomenclature would help to indicate that the item will only be added if it doesn't already exist, e.g. add_unique?

Reply:
Or maybe `insert`?

Reply:
I renamed `push` to `insert` and `append` to `extend`. If you prefer something else, let me know.